### PR TITLE
Fix alignMols so that it takes into account of QueryAtoms and QueryBonds

### PR DESCRIPTION
**Pull request by DoliathGavid**
This is a quick fix so that the substructure matcher in the align tools takes into account of QueryAtoms and QueryBonds.

This may introduce backwards incompatibility issues if there's alignment code that for some reason depended on the lack of query matching (which albeit, is a bit weird).

### CODE REVIEW

File: Code/GraphMol/MolAlign/AlignMolecules.cpp [hunk 1]
```diff
@@ -30,7 +30,7 @@ double getAlignmentTransform(const ROMol &prbMol, const ROMol &refMol,
   if (atomMap == 0) {
     // we have to figure out the mapping between the two molecule
     MatchVectType match;
-    if (SubstructMatch(refMol, prbMol, match)) {
+    if (SubstructMatch(refMol, prbMol, match, true, false, true)) {
```
Comment (bp-kelley) [inline Code/GraphMol/MolAlign/AlignMolecules.cpp]: In general I don't like magic numbers (or magic booleans in this case).  It makes it easier on future devs to be explicit and also helps validate the pull request :)

```
const bool RecursionPossible=true;
const bool UseChirality=false;
const bool UseQueryQueryMatches=true;
if (SubstructMatch(ref mol, prbMol, match, 
                              RecursionPossible, UseChirality, UseQueryQueryMatches))
```

**Comment by bp-kelley** [comment]
At the minimum this change will require a regression test and perhaps an example in the documentation of how one can align queries to queries.

**Comment by DoliathGavid** [comment]
+1 on getting rid of magic booleans. It's not quite clear to me how to write a regression test for this yet, any suggestions?

**Comment by greglandrum** [comment]
Maybe use one of the existing tests of `alignMol()` from `Code/GraphMol/MolAlign/testMolAlign.cpp` where you've updated the input structures to include queries?

**Comment by DoliathGavid** [comment]
After digging around, this is isn't a totally safe change. I was hoping that this would always result in alignment identifying more atoms to use for the alignment. However, because the semantics of SubstructMatch changes, this could break backwards compatibility for people who are setting atomic numbers and other properties on the base Atom class of QueryAtoms (in addition to setting various QueryAtom properties).

Nevertheless, I think for the purpose of aligning molecules, this change still makes sense.

None of this would be a problem if we forbade the base Atom class of QueryAtoms actually representing real atoms of any sort. (Eg. completely disallow QueryAtoms from having non zero AtomicNumbers). 

```
  std::string rdbase = getenv("RDBASE");
  std::string fname1 = rdbase + "/Code/GraphMol/MolAlign/test_data/1oir.mol";
  RWMol *m1 = new RWMol(*MolFileToMol(fname1));
  std::string fname2 =
      rdbase + "/Code/GraphMol/MolAlign/test_data/1oir_conf.mol";
  RWMol *m2 = new RWMol(*MolFileToMol(fname2));

  // atomCompat never triggers QueryMatch() if only one of the two atoms is a query
  
  auto a1 = new QueryAtom();
  auto a2 = new QueryAtom();

  // UNCHANGED
  a1->setAtomicNum(5);
  a2->setAtomicNum(5);
  a1->setQuery(makeAtomNumQuery(7));
  a2->setQuery(makeAtomNumQuery(8));
  m1->replaceAtom(28, a1);
  m2->replaceAtom(19, a2);

  MatchVectType match;

  std::cout << "BEFORE" << SubstructMatch(*m1, *m2, match, true, false, false) << std::endl; // 0
  std::cout << "AFTER" << SubstructMatch(*m1, *m2, match, true, false, true) << std::endl; // 0
    
  // IMPROVED
  auto a3 = new QueryAtom();
  auto a4 = new QueryAtom();
  a3->setAtomicNum(0);
  a4->setAtomicNum(0);
  a3->setQuery(makeAtomNumQuery(7));
  a4->setQuery(makeAtomNumQuery(7));

  m1->replaceAtom(28, a3);
  m2->replaceAtom(19, a4);

  std::cout << "BEFORE" << SubstructMatch(*m1, *m2, match, true, false, false) << std::endl; // 0
  std::cout << "AFTER" << SubstructMatch(*m1, *m2, match, true, false, true) << std::endl; // 1

  // UNCHANGED
  auto a5 = new QueryAtom();
  auto a6 = new QueryAtom();
  a5->setAtomicNum(7);
  a6->setAtomicNum(7);
  a5->setQuery(makeAtomNumQuery(7));
  a6->setQuery(makeAtomNumQuery(7));

  m1->replaceAtom(28, a5);
  m2->replaceAtom(19, a6);

  std::cout << "BEFORE" << SubstructMatch(*m1, *m2, match, true, false, false) << std::endl; // 1
  std::cout << "AFTER" << SubstructMatch(*m1, *m2, match, true, false, true) << std::endl; // 1

  // REGRESSION
  auto a7 = new QueryAtom();
  auto a8 = new QueryAtom();
  a7->setAtomicNum(7);
  a8->setAtomicNum(7);
  a7->setQuery(makeAtomNumQuery(7));
  a8->setQuery(makeAtomAliphaticQuery());

  m1->replaceAtom(28, a7);
  m2->replaceAtom(19, a8);

  std::cout << "BEFORE" << SubstructMatch(*m1, *m2, match, true, false, false) << std::endl; // 1
  std::cout << "AFTER" << SubstructMatch(*m1, *m2, match, true, false, true) << std::endl; // 0

  return;
```

**Comment by DoliathGavid** [comment]
So one solution is to combined QueryAtom::Match and QueryAtom::QueryMatch into a single proper polymorphic function:

```cpp
bool QueryAtom::Match(Atom const *what) const {
  PRECONDITION(what, "bad query atom");
  PRECONDITION(dp_query, "no query set");
  return dp_query->Match(what);
}
```

to

```cpp
bool QueryAtom::Match(Atom const *what) const {
  PRECONDITION(what, "bad query atom");
  PRECONDITION(dp_query, "no query set");
  if(what->hasQuery()) {
    return queriesMatch(dp_query, what->getQuery());
  } else {
    return dp_query->Match(what);
  }
}
```

**Comment by DoliathGavid** [comment]
I'm sure you already know this, but I *just* realized that Match isn't even symmetric:
```
  auto a0 = new Atom(6);
  auto a1 = new QueryAtom(1);
  a1->setQuery(makeAtomNumQuery(6));

  std::cout << a1->Match(a0) << std::endl; // 1
  std::cout << a0->Match(a1) << std::endl; // 0
```

**Comment by bp-kelley** [comment]
Match should not be symmetric, as query atoms aren't really atoms.  It could quite complicated indeed in the general case (think recursive smarts, or matchers etc).  For a much better discussion see here: https://www.mail-archive.com/rdkit-discuss@lists.sourceforge.net/msg03883.html

And actually, you can replace query atoms with fairly arbitrary code, i.e. match two atoms if they have the same property tag.

In a pinch you could add it as an option (although the function is pretty top heavy already)

```
double getAlignmentTransform(const ROMol &prbMol, const ROMol &refMol,
                             RDGeom::Transform3D &trans, int prbCid, int refCid,
                             const MatchVectType *atomMap,
                             const RDNumeric::DoubleVector *weights,
                             bool reflect, unsigned int maxIterations,
                             bool UseQueryQueryMatches=false
) 
```

Without a concrete example of the issues you are trying to solve, I worry that this might be a solution looking for a problem.  I can completely understand adding query atoms to the probe molecule to better align more targets, but I expect you may be thinking that query query matches allow more things to match which is probably not true, it might actually make fewer match because query query matches have to be exact.

**Comment by DoliathGavid** [comment]
It's not the non-symmetry itself that bothers me, but the part (as mentioned in the linked thread) whereby a normal atom's properties are matched against the BASE class of a query atom. I (again this is just me) *always* set all the properties of the query atom to zero values via the default constructor, and only customize the query property. So I'd *always* want to call queryatom's match on the normal atom instead. I'd ideally want Match() to behave as follows:

| this/what        | QueryAtom           | Atom  |
| ------------- |:-------------:| -----:|
| QueryAtom      | queryquerymatch | querymatch |
| Atom     | querymatch      |  match |

instead of:

Match():

| this/what        | QueryAtom           | Atom  |
| ------------- |:-------------:| -----:|
| QueryAtom      | querymatch | querymatch |
| Atom     | match      |  match |

QueryMatch():

| this/what        | QueryAtom           | Atom  |
| ------------- |:-------------:| -----:|
| QueryAtom      | queryquerymatch | querymatch |

This behavior would be also be quite easy to implement.

PS: The use case is that I don't have a clear cut delineation between query and target when mols start to have both QueryAtoms and normal Atoms attached to it.

**Comment by bp-kelley** [comment]
I think there are distinct reasons not to have this as the default behavior of the default query matching.

I admit to being confused about what you are asking, it really doesn't match with my experience (pardon the pun, it's late).  I tend to do something along the lines of (also apologies for python code here, but it's easier to test), but isn't what you want exactly what replaceAtom is for?  In the code below I'm changing nitrogens in m2 into atoms that look for properties with certain values, there is no matching of anything but the query:

```
from rdkit import Chem
from rdkit.Chem import rdqueries

m = Chem.MolFromSmiles("CCCCC")
m2 = Chem.RWMol(Chem.MolFromSmiles("NN"))

for atom in m2.GetAtoms():
    m2.ReplaceAtom(atom.GetIdx(),
                  rdqueries.HasIntPropWithValueQueryAtom("bar", atom.GetIdx()))

print "done"
assert m.HasSubstructMatch(m2)==False

# set the props so we can now match
m.GetAtomWithIdx(0).SetProp("bar", "0")
m.GetAtomWithIdx(1).SetProp("bar", "1")

assert m.HasSubstructMatch(m2)==True
```
However, the symmetry is now quite broken:  "m2.HasSubstructMatch(m) == False" which I think is what we are talking about here.

The query molecule stored as a psuedo-real molecule is a debate that I have had with Greg before and we ended up on different sides of the fence.  I think that it can seem like the matches should be symmetric, however in the previous toolkits I have been involved with, the use or a smarts pattern was similar to generating regular expressions:

```
pat = smarts.compile("[C,N]")
groups = pat.match( molecule )
```

One wouldn't necessarily expect this to be symmetric as the matcher is a finite state machine and the molecule is not :)  However, I will concede that adding symmetry might be useful, but I'm not sure that this is what QueryQueryMatches actually do.  I expect what you want is something like:

```
  match = probe_atom->Match(refatom) || refatom->Match(probeatom);
  and likewise for bonds.
```

Here is the atom matching at the bottom level (Code/GraphMol/Atom.cpp)

```
bool QueryAtom::Match(Atom const *what) const {
  PRECONDITION(what, "bad query atom");
  PRECONDITION(dp_query, "no query set");
  return dp_query->Match(what);
}
bool QueryAtom::QueryMatch(QueryAtom const *what) const {
  PRECONDITION(what, "bad query atom");
  PRECONDITION(dp_query, "no query set");
  if (!what->hasQuery()) {
    return dp_query->Match(what);
  } else {
    return queriesMatch(dp_query, what->getQuery());
  }
}
```

And QueryQueryMatch is pretty much a straight up do the two query atom match exactly.  This note has gotten long enough that I might be completely on the wrong track, so please let me know if I am.

**Comment by DoliathGavid** [comment]
>> I admit to being confused about what you are asking, it really doesn't match with my experience (pardon the pun, it's late). Isn't what you want exactly what replaceAtom is for?

My point is that for me personally, there's no point to QueryAtom subclassing Atom save for the fact that it lets us attach it to Mol objects easily. QueryAtom doesn't  really use any of the base class members (eg: df_isAromatic, d_atomicNum, d_chiralTag etc.). To me, it's an entirely ethereal construct whose primary purpose is for querying, not representing real atoms. As such, to have a regular Atom::Match(what) compare against properties of this ethereal QueryAtom's base class is just plain weird to me.

>> However, the symmetry is now quite broken: "m2.HasSubstructMatch(m) == False" which I think is what we are talking about here.

>> However in the previous toolkits I have been involved with, the use or a smarts pattern was similar to generating regular expressions.

I view the SMARTS as nothing more than encoding of query mols analogous to SMILES as being an encoding of normal molecules. But I can certainly see the analogy to FSMs.

>>   I expect what you want is something like:  match = probe_atom->Match(refatom) || refatom->Match(probeatom);

I have something much more fun:

```
bool Atom::Match(Atom const *what) const {
  PRECONDITION(what, "bad query atom");

  // this is needed to symmetrize Matching
  if(dynamic_cast<const QueryAtom *>(what) && what->hasQuery()) {
    return what->Match(this);
  }
```

**Comment by bp-kelley** [comment]
If it is just the subclass issue, I think that is because QueryAtom adds extra data to the atom (in this case a pointer to a query structure) that isn't needed in general, so why use the extra memory hit.  There are a lot of atoms in general.  One could argue that this could be moved into a property, but I think that is the main reason.  If we assume this is the case, a QueryAtom has to be an Atom since that is how the graph is traversed.  You are correct that any additional atom information is essentially lost in the query, but I think that is a small price to pay here.

Also note that you don't have to dynamic cast, you can just all _what->hasQuery()_.  Your Match proposal is along the lines I was thinking.   Now the question is what are the ramifications of this?  I worry with recursive smarts the semantics of "SubstructureMatch" won't make sense mainly because switching the matching order may be doing a "SuperstructureMatch" instead.

File: Code/GraphMol/MolAlign/testMolAlign.cpp [hunk 2]
```diff
@@ -66,6 +66,58 @@ void test1MolAlign() {
   delete m3;
 }
 
+void test1MolWithQueryAlign() {
+  // identical to test1MolAlign except we replace one atom with a QueryAtom instead
+
+  std::string rdbase = getenv("RDBASE");
+  std::string fname1 = rdbase + "/Code/GraphMol/MolAlign/test_data/1oir.mol";
+  RWMol *m1 = new RWMol(*MolFileToMol(fname1));
+  auto a1 = new QueryAtom(6);
```
Comment (bp-kelley) [inline Code/GraphMol/MolAlign/testMolAlign.cpp]: We're not using C++11 yet.

**Comment by DoliathGavid** [comment]
>> If it is just the subclass issue, I think that is because QueryAtom adds extra data to the atom (in this case a pointer to a query structure) that isn't needed in general, so why use the extra memory hit.

It's more a matter of QueryAtom subclassing Atom. I'd actually propose both QueryAtom and Atom subclassing a generic Node class instead :) but that's opening another can of worms I won't get into for now. 

>>  I worry with recursive smarts the semantics of "SubstructureMatch" won't make sense mainly because switching the matching order may be doing a "SuperstructureMatch" instead.

I guess it comes down to how we view Matching: either as a subset operator or as an order-invariant, symmetric compatibility operator. I'm arguing for the latter because Atoms aren't really proper subsets of QueryAtoms to begin with, since all the dp_* properties in the base class are just dummy attributes to start with, and they're only there because it's convenient for attaching to Mols.

As for the actual PR, I totally forgot we weren't using C++11 yet. I also added explicit names to the magic booleans. I agree that alignment transforms are top-heavy to begin with, and for all the reasons we listed above, I think it's *ok* to replace the default behavior. I'll do a quick rebase to clean up once you or greg think its ready to go.

**Comment by DoliathGavid** [comment]
Hm.. python 2.7 builds failed on building the java wrappers, but python 3.x passed and so did appveyor
```
1 error
5 warnings
Generating /home/travis/build/rdkit/rdkit/Code/JavaWrappers/gmwrapper/doc/index.html...
Generating /home/travis/build/rdkit/rdkit/Code/JavaWrappers/gmwrapper/doc/help-doc.html...
Note: Custom tags that could override future standard tags:  @example, @notes. To avoid potential overrides, use at least one period character (.) in custom tag names.
5 warnings
make[2]: *** [../Code/JavaWrappers/gmwrapper/org.RDKit.jar] Error 1
make[2]: *** Deleting file `../Code/JavaWrappers/gmwrapper/org.RDKit.jar'
make[1]: *** [Code/JavaWrappers/gmwrapper/CMakeFiles/GraphMolWrapJar.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
[ 50%] [34m[1mGenerating ../../../../Code/JavaWrappers/gmwrapper/build-test/org/RDKit/WrapperTests.class[0m
/home/travis/build/rdkit/rdkit/Code/JavaWrappers/gmwrapper/src-test/org/RDKit/GraphMolTest.java:67: error: cannot find symbol
	public static void assertPointEquals(Point3D pt1, Point3D pt2, double tol) {
	                                     ^
  symbol:   class Point3D
  location: class GraphMolTest
/home/travis/build/rdkit/rdkit/Code/JavaWrappers/gmwrapper/src-test/org/RDKit/GraphMolTest.java:67: error: cannot find symbol
	public static void assertPointEquals(Point3D pt1, Point3D pt2, double tol) {
```

**Comment by bp-kelley** [comment]
@DoliathGavid sometimes this happens, I usually just restart the travis build. (the circle button with the arrow).

While I'm not in favor of a node subclass per-se, I am in favor of an abstract API that can be fulfilled by an implementation class.  What goes into that abstract API should be atom specific though, so we probably will disagree on that point.

What might be amusing is that we actually do have a node API in the underlying boost-graph, so there are actually two ways of getting the underlying graph structure which I personally think if overkill, though some would say that is the best kind of kill.

I'm still waffling on the symmetric match, I can see it from an atom perspective, but not from a substructmatch perspective, but perhaps I am overthinking this.  It boils down to this:

subtract match query qatom -> Carbon ( that must be in an an enol )
target atom -> carbon that is in an enol

Are these really symmetric?  Should they be?

**Comment by DoliathGavid** [comment]
OK it looks like the tests pass.

Maybe we can defer the symmetry problem to a later discussion?

**Comment by DoliathGavid** [comment]
Is there anything else that needs to be done to punt on this immediate PR?

**Comment by greglandrum** [comment]
Yeah, my day needs to get about 4 hours longer. :-S

I'll try to make some time to review it later today.

**Comment by greglandrum** [comment]
To be clear before I really get into this:
- I consider the fact that an ROMol can contain Atoms and QueryAtoms to be a good thing. I believe it makes the toolkit simpler and more natural to use. It may be different from how other toolkits handle things, but that isn't something I find remotely alarming.
- Queries are not, and don't need to be, symmetric. The operator is "Match", not "Equals". You don't even need to look to QueryAtoms to see that this is the case:
```
In [16]: a1 = Chem.Atom(6)

In [17]: a3 = Chem.Atom(6)

In [18]: a3.SetFormalCharge(1)

In [19]: a1.Match(a3)
Out[19]: True

In [20]: a3.Match(a1)
Out[20]: False
```
This behavior is documented here: http://www.rdkit.org/docs/RDKit_Book.html#atom-atom-matching-in-substructure-queries

Changing either of these things would require a fairly large amount of RDKit code to be rewritten. It's a backwards incompatible change, so it would break a completely unknown amount of code that other people have written that uses the RDKit, and would deliver, as far as I can tell, no actual value.

File: Code/GraphMol/MolAlign/AlignMolecules.cpp [hunk 3]
```diff
@@ -30,7 +30,10 @@ double getAlignmentTransform(const ROMol &prbMol, const ROMol &refMol,
   if (atomMap == 0) {
     // we have to figure out the mapping between the two molecule
     MatchVectType match;
-    if (SubstructMatch(refMol, prbMol, match)) {
+    const bool recursionPossible = true;
```
Comment (greglandrum) [inline Code/GraphMol/MolAlign/AlignMolecules.cpp]: This set of changes is backwards incompatible.
If you want control over these matching options, feel free to add optional arguments to the function.

File: Code/GraphMol/MolAlign/AlignMolecules.cpp [hunk 4]
```diff
@@ -30,7 +30,10 @@ double getAlignmentTransform(const ROMol &prbMol, const ROMol &refMol,
...
+    const bool useChirality = false;
+    const bool useQueryQueryMatches = true;
```
Comment (greglandrum) [inline Code/GraphMol/MolAlign/AlignMolecules.cpp]: This piece isn't backwards compatible, but it's enough of an edge case that I think it's unlikely to be a problem.

**Comment by greglandrum** [comment]
This looks fine now.
Thanks for submitting the PR!

**Pull request merged**